### PR TITLE
Trial Banner

### LINKED ISSUE #48522: Banner for instance in trial
__Context__
We want to nudge our cloud customers on trial to activate their instance.

__Problem__
We do not have a mechanism for us to show our customers that their trial will soon end.

__Proposal__

[Product doc](https://www.notion.so/metabase/No-CC-required-for-Cloud-trial-d268a1a7f1d94990aef261cf1211aa76?pvs=4#10e69354c901800a802cd38e56aedd5b).

- [x] We need to display a banner for to admin users of an instance in trial, similarly to the past-due payment banner. 
- [x]  Banner appears 14d before trial ends
- [x]  X dismisses the banner
- [x]  banner appears 3d before trial ends, 2d before trial ends, 1d before trial ends (this is just conditional on the days until the end of trial, does not check if the banner was dismissed or not)
- [ ] Depends on https://github.com/metabase/infra-frontend/issues/1153, exact deep link url tbd 

[Figma with the updated design](https://www.figma.com/design/OGkkIYPpfcEfd7EjnmL6SJ/No-CC-required-for-Cloud-trial?node-id=657-1054&t=8zdpNzGRg0uz7WJ0-4)

Image

### PR DESCRIPTION
**Pull request by nemanjaglumac**
### What does this PR accomplish?
Resolves https://github.com/metabase/metabase/issues/48522.

tl;dr
- We're adding a new app banner that's **only visible and applicable to hosted instances in trial mode**.
    - It should navigate users to the page in store where they can easily manage their subscription.
    - This is a part of the larger "no credit card" sign up initiative

The logic for the banner (re) appearance is as follows:
1. Up until three days before the trial expires it is enough that a user dismisses this banner
2. On the three day before token expiry threshold (mark), the banner re-appears and needs to be dismissed again
3. The same thing happens on the two day mark, one day mark and finally on the 0 day (last 24h) mark

image

## Tests
This feature is covered with a lot of unit tests. I plan to add an E2E test in a separate PR. It will involve some mocking and probably playing with the `cy.clock()` so I really don't want to (1) block this PR any longer and (2) increase the number of changed files for now. cc @iethree

### CODE REVIEW

File: frontend/src/metabase/components/AppBanner/AppBanner.unit.spec.tsx [hunk 1]
```diff
@@ -152,4 +158,168 @@ describe("AppBanner", () => {
...
+      trial: true,
+    };
+
+    const getTokenExpiryDate = (daysRemaining: number) => {
+      return (
+        dayjs()
+          .add(daysRemaining, "days")
+          // Safety buffer - even a second would do but we want to
+          // account for the slower testing environments, like CI.
+          .add(1, "minute")
```
Comment (nemanjaglumac) [inline frontend/src/metabase/components/AppBanner/AppBanner.unit.spec.tsx:174]: The alternative would be to mock `now` in order to "freeze the time". Either way is fine with me.

File: frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  if (!shouldShowBanner) {
+    return null;
+  }
+
+  return (
+    <Banner
+      icon="warning_round_filled"
+      // This is --mb-base-color-dubloon-10, which is not allowed to be referenced directly.
+      // FIXME: Update the next line once we switch to semantic colors.
+      bg="hsla(46, 96%, 90%, 1)"
```
[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:58]: Yikes! please don't do this. use `--mb-color-warning` (which you can use in mantine with just "warning") - which we want to be uniform throughout the app, and hopefully one day soon will be more dubloony. If this makes your local designer unhappy, send them to me and I will buy them a turtleneck or something.

[the remark this answers is not in the document]
Comment (nemanjaglumac) [reply frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx]: Done.

https://github.com/metabase/metabase/pull/49344/commits/d24f147c576edcddfe18ce01ec5eecc7887599f4

[the remark this answers is not in the document]
Comment (trinya) [reply frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx]: @iethree can we make the --mb-color-warning less yellow? where is it used except from this banner? we can have a slack thread if it helps.

[the remark this answers is not in the document]
Comment (nemanjaglumac) [reply frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx]: We can't.
There is a predefined set of colors currently available in the main app. We absolutely cannot alter the color for a single use case.

The design team came up with a nice color system, but we are currently unable to use it to its full extent. That will change once we upgrade to Mantine v7.

File: frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+      const bannerReappearanceThreshold = tokenExpiryDate
+        .subtract(daysRemaining, "day")
+        .unix();
+
+      // Banner shows if it hasn't been dismissed after the reappearance point
+      return !lastDismissed || lastDismissed < bannerReappearanceThreshold;
+    }
+
+    // For the first 11 days, check if it was ever dismissed
+    return !lastDismissed;
```
Comment (iethree) [inline frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:40]: The contents of this `useMemo` should be a separate function in a `utils` file with the signature:

```ts
function shouldShowBanner(tokenExpiryDate, lastDismissed)
```

That way you can test nearly all the complex logic in this component with very simple unit tests without the need for mocking settings and whatnot.

File: frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+import { getStoreUrl } from "metabase/selectors/settings";
+import { Flex, Text } from "metabase/ui";
+import type { TokenStatus } from "metabase-types/api";
+export const TrialBanner = ({ tokenStatus }: { tokenStatus: TokenStatus }) => {
+  const [lastDismissed, setLastDismissed] = useUserSetting(
+    "trial-banner-dismissal-timestamp",
+  );
+
+  const now = dayjs();
+  const tokenExpiryDate = dayjs(tokenStatus["valid-thru"]);
```
Comment (iethree) [inline frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:18]: days() is going to get a date from the browser in the user's timezone. I think/guess that our tokens are storing timestamps in UTC? this could create some weird edge cases. We should have tests covering this. Hooray timezones!

File: frontend/src/metabase/components/AppBanner/AppBanner.unit.spec.tsx [hunk 5]
```diff
@@ -97,8 +100,11 @@ describe("AppBanner", () => {
     it("should render an error with details when the token is `invalid`", () => {
       setup({
         isAdmin: true,
-        tokenStatusStatus: "invalid",
-        tokenError: "This is a critical damage.",
+        tokenStatus: {
+          ...token,
+          status: "invalid",
+          "error-details": "This is a critical damage.",
```
Comment (iethree) [inline frontend/src/metabase/components/AppBanner/AppBanner.unit.spec.tsx:106]: 🙀

File: frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+
+    // For the first 11 days, check if it was ever dismissed
+    return !lastDismissed;
+  }, [daysRemaining, lastDismissed, tokenExpiryDate]);
+
+  const href = getStoreUrl("account/manage/plans");
+
+  const handleBannerClose = () => {
+    setLastDismissed(dayjs().unix());
+  };
```
Comment (iethree) [inline frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:47]: nit, but I would put these below the early return. No need to calculate them if we're not using them.

File: frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+    if (daysRemaining <= 3) {
+      const bannerReappearanceThreshold = tokenExpiryDate
+        .subtract(daysRemaining, "day")
+        .unix();
+
+      // Banner shows if it hasn't been dismissed after the reappearance point
+      return !lastDismissed || lastDismissed < bannerReappearanceThreshold;
+    }
+
+    // For the first 11 days, check if it was ever dismissed
```
Comment (iethree) [inline frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:39]: Bug: there is no logic in the code that I can see that prevents this from triggering more than 11 days out. I'm pretty sure in its current state, every trial customer will see a notification that their trial will expire in X days, even if X is 90. The tests also don't cover any number of days greater than 11

**Review by iethree — CHANGES_REQUESTED** [review]
Nice job on this feature! ❤️ 

Found one bug, which could really benefit from a little refactoring to make the core date logic easier to test

File: frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    // No banner if the trial already expired
+    if (daysRemaining < 0) {
+      return false;
+    }
+
+    // In the last 3 days, check dismissal logic with stricter reappearance
+    if (daysRemaining <= 3) {
+      const bannerReappearanceThreshold = tokenExpiryDate
+        .subtract(daysRemaining, "day")
+        .unix();
```
Comment (iethree) [inline frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:33]: This is pretty hard to follow. I think what this is saying is that "if the token expires in 3 days, we want to check if the last banner dismissal was before 3 days before the trial expires" ?  🥴 

Maybe instead, let's just check if the banner was dismissed today?


```ts
const wasDismissedToday = dayjs().format("YYYYMMDD") === dayjs(lastDismissed).format("YYYYMMDD");

return !wasDismissedToday;
```

Comment (nemanjaglumac) [reply frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:33]: Genial!
I was so deep into this problem/code that I missed the forest for the trees. This is way simpler. Thank you.

File: frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx (hunk 7 above)
Comment (nemanjaglumac) [reply frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx:39]: No bug, but very vague product documentation.
I double checked with the PM, and the verdict is - as soon as the trial starts (no matter how long it is), show the banner. I'll update tests to make this more explicit.

File: frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx (hunk 4 above)
Comment (nemanjaglumac) [reply frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx]: This is great observation. It sent me on a long journey, and I thank you for it.
Ironically, it turned out that dayjs automatically handles timezone differences in its methods like `.isSame()`, `.diff()`, etc.

However, I decided to still handle this in the code (1) for the sake of explicitness, and (2) for the sake of easier testing. After all, we might want to change/replace this library with something else in the future. This ensures we don't forget to handle the timezone offset.

Thank you!

File: frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx (hunk 7 above)
Comment (iethree) [reply frontend/src/metabase/nav/components/TrialBanner/TrialBanner.tsx]: 🥴

File: frontend/src/metabase/components/AppBanner/AppBanner.tsx [hunk 9]
```diff
@@ -28,6 +40,28 @@ export const AppBanner = () => {
...
+    });
+
+    return (
+      showBanner && (
+        <TrialBanner
+          daysRemaining={daysRemaining}
+          onClose={() => setLastDismissed(getCurrentUTCTimestamp())}
+        />
+      )
+    );
```
Comment (iethree) [inline frontend/src/metabase/components/AppBanner/AppBanner.tsx:62]: I think you want to put this return statement in another `if` statement. This logic short-circuits the possibility of showing the payment banner to any trial users.. and is harder to understand than just falling through to the `null` return at the end of the function

```suggestion
    if(showBanner) {
        return (
          <TrialBanner
            daysRemaining={daysRemaining}
            onClose={() => setLastDismissed(getCurrentUTCTimestamp())}
          />
      );
    }

```

**Review by iethree — APPROVED** [review]
Looks good! take a look at these few nits before merging 👇

File: frontend/src/metabase/components/AppBanner/AppBanner.tsx (hunk 9 above)
Comment (nemanjaglumac) [reply frontend/src/metabase/components/AppBanner/AppBanner.tsx:62]: Payment banner and Trial banner are mutually exclusive. One is shown to the self-hosted instances and the other one is shown to the hosted/cloud instances.

But I can still reorder this if it makes it easier to read.

File: frontend/src/metabase/components/AppBanner/utils.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+ *
+ * If the trial is within the last 3 days, the function checks if the banner
+ * was dismissed on the current day. Unless it was dismissed on that day,
+ * the banner will be shown to the user.
+ *
+ * If the trial has more than 3 days remaining, the banner will be shown only if
+ * the user has never dismissed it.
+ */
+export function shouldShowTrialBanner({
+  daysRemaining,
```
[the remark this answers is not in the document]
Comment (nemanjaglumac) [reply frontend/src/metabase/components/AppBanner/utils.ts:29]: Honestly, this decision wasn't so much about the performance as much as it was about making this helper pure. I really really **really** didn't want to compute the days remaining inside the function, which makes it dependent on `now` (or the currentTimestamp). Maybe it's just me, but this decision actually made tests much more simple.

If you have two ISO strings as an input, all calculations you do with them are always going to yield the same results. Which also answers to [this comment](https://github.com/metabase/metabase/pull/49344#discussion_r1868408656).

File: frontend/src/metabase/components/AppBanner/utils.ts [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+  }
+
+  // Show banner if it was never dismissed
+  if (!lastDismissed) {
+    return true;
+  }
+
+  // In the last 3 days, check that the banner was dismissed that day
+  if (daysRemaining <= 3) {
+    const today = dayjs(tokenExpiryTimestamp).subtract(daysRemaining, "days");
```
[the remark this answers is not in the document]
Comment (nemanjaglumac) [reply frontend/src/metabase/components/AppBanner/utils.ts:45]: Please see https://github.com/metabase/metabase/pull/49344#discussion_r1869312211

**Comment by nemanjaglumac** [comment]
@metabase-bot backport release-x.51.x
